Intersect.Server/Entities/Player.cs

```diff
@@ -3518,88 +3518,76 @@ public void SellItem(int slot, int amount)
-                                {
-                                    TryTakeItem(
-                                        FindInventoryItemSlot(
-                                            shop.SellingItems[slot].CostItemId,
-                                            shop.SellingItems[slot].CostItemQuantity * buyItemAmt
-                                        ), shop.SellingItems[slot].CostItemQuantity * buyItemAmt
-                                    );
+            var currencyBase = ItemBase.Get(boughtItem.CostItemId);
+            if (currencyBase == default)
+            {
+                PacketSender.SendChatMsg(this, Strings.Shops.error, ChatMessageType.Inventory, CustomColors.Alerts.Error, Name);
+                return;
+            }
+
+            var boughtItemAmount = 1;
+            if (boughtItemBase.IsStackable)
+            {
+                boughtItemAmount = Math.Max(1, amount);
+            }
 
-                                    TryGiveItem(buyItemNum, buyItemAmt);
+            var currencyAmount = FindInventoryItemQuantity(currencyBase.Id);
+            var itemCostTotal = boughtItem.CostItemQuantity * boughtItemAmount;
+            if (currencyAmount < itemCostTotal)
+            {
+                PacketSender.SendChatMsg(this, Strings.Shops.cantafford, ChatMessageType.Inventory, CustomColors.Alerts.Error, Name);
+                return;
+            }
 
-                                    if (!TextUtils.IsNone(shop.BuySound))
-                                    {
-                                        PacketSender.SendPlaySound(this, shop.BuySound);
-                                    }
-                                }
-                                else
-                                {
-                                    PacketSender.SendChatMsg(
-                                        this, Strings.Shops.inventoryfull, ChatMessageType.Inventory, CustomColors.Alerts.Error, Name
-                                    );
-                                }
-                            }
-                        }
-                        else
-                        {
-                            PacketSender.SendChatMsg(this, Strings.Shops.cantafford, ChatMessageType.Inventory, CustomColors.Alerts.Error, Name);
-                        }
+            if (!CanGiveItem(boughtItemBase.Id, boughtItemAmount))
+            {
+                PacketSender.SendChatMsg(this, Strings.Shops.inventoryfull, ChatMessageType.Inventory, CustomColors.Alerts.Error, Name);
+            }
+
+            if (itemCostTotal > 0)
+            {
+                var currencySlots = FindInventoryItemSlots(currencyBase.Id);
+                int remainingCost = itemCostTotal;
+
+                foreach (var itemSlot in currencySlots)
+                {
+                    int quantityToRemove = Math.Min(remainingCost, itemSlot.Quantity);
+                    TryTakeItem(itemSlot.ItemId, quantityToRemove);
```

Comment:
Honestly if TryTakeItem fails it needs to roll-back all of the items it already took, and it needs to abort the for loop and show a message. It should not assume that it succeeded.

Reply:
It should also log what was successfully taken, what failed to be taken, what the quantityToRemove was when it failed, and the player ID it failed on.

Reply:
This call should also use TryTakeItem(itemSlot) instead of itemSlot.ItemId

Reply:
@WeylonSantana , following what @lodicolo said, try something like this perhaps:

```csharp
if (itemCostTotal > 0)
{
    var currencySlots = FindInventoryItemSlots(currencyBase.Id);
    int remainingCost = itemCostTotal;
    bool success = true;

    foreach (var itemSlot in currencySlots)
    {
        int quantityToRemove = Math.Min(remainingCost, itemSlot.Quantity);
        if (TryTakeItem(itemSlot, quantityToRemove))
        {
            remainingCost -= quantityToRemove;
            Log.Warn($"Successfully took {quantityToRemove} items from slot {itemSlot} for player {playerId}.");
        }
        else
        {
            success = false;
            Log.Warn($"Failed to take items from slot {itemSlot} for player {playerId}. Quantity to remove: {quantityToRemove}.");
            break;
        }

        if (remainingCost <= 0)
        {
            break;
        }
    }

    if (!success)
    {
        // TODO: Rollback code goes here ?
        Log.Warn("Rolling back items due to failed transaction.");
    }
}
```

could *probably* do something similar with the giving item case:
 https://github.com/AscensionGameDev/Intersect-Engine/pull/1832/files#r1264234149
 
 PS. this is just an idea/example, remember use Localization/Strings.cs for the strings here!
 
 ie: 
```csharp
[JsonProperty(NullValueHandling = NullValueHandling.Ignore)]
public readonly LocalizedString GiveItemSuccess = "Successfully took {0} items from slot {1} for player {2}.";
```